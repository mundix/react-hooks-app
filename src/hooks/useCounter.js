import { useState } from 'react';
//un custom hook no es mas que una simple function 

export const useCounter = (initialState = 10) => {
    
    const [counter, setCounter] = useState( initialState);

    const increment = () => {
    // const increment = (factor = 1) => {
        setState( state + 1);
        // setState( state + factor);
    }

    // const decrement = (factor = 1) => {
    const decrement = () => {
        setState( state - 1);
        // setState( state - factor);
    }

    const reset = () => {
        setState(initialState);
    }

    // se peude refresar un objeto por si necesito incrementar o decrementar 
    return  {
        counter,
        increment,
        decrement,
        reset
    }

}