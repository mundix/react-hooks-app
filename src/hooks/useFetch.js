import { useEffect, useState } from "react"

export const useFetch = ( url ) => {
     
    // Se puede poenralguna condicion al url si no viene 
    const [state, setState] = useState({ data: null, loading: true, error: null});

    useEffect( () => {

        setState({ data: null, loading: true, error: null}); //regresa un nuevo stado en el estado inicial.

        fetch( url)
        .then( resp => resp.json())
        .then( data => {
            setState({
                loading: false,
                error: null,
                data 
            })
        });
    }, [url]);

    return state;
}
