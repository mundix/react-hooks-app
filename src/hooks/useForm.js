import { useState } from "react"

export const useForm = (initialState = {}) => {
    const [values, setValues] = useState(initialState);

    const handleInputChange = ({target}) => {
        setValues({
             ...values,
             [target.name] : target.value
         });
    }
    //Devolvere un arraay con los valores de setValue, y el metodo 
    return [values, handleInputChange];
}
