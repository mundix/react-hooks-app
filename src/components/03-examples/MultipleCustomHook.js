import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import '../02-useEffect/effect.css';

export const MultipleCustomHook = () => {

    const state = useFetch('https://breakingbadapi.com/api/quotes');
    

    console.log(state);

    return (
        <div>
            <h1>Custom Hooks!!!</h1>
        </div>
    )
}
