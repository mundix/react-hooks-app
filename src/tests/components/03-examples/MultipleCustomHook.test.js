import React from 'react';
import {shallow} from 'enzyme';
import { MultipleCustomHook } from '../../../components/03-examples/MultipleCustomHook';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

// Esto lo que hace Cuando vaya a utilizar este artchivo , en lugar de usar el useFetch  , va a utilizar una implementacion del useFetch
jest.mock('../../../hooks/useFetch');
// jest.mock('../../../hooks/useCounter');


describe('Pruebas en <MultipleCustomHook/>', () => {
    
    // Ahora quiero definir como trabeje el useCounter 
    // useCounter.mockReturnValue({
    //     counter: 10,
    //     increment: () => {} //simplemente defino la funcion para que no haga nada
    // });


    test('debe mostrase correctamente', () => {

        //cuando se llame el useFetch
        useFetch.mockReturnValue({
            data:null,
            loading: true,
            error:null
        });

        const wrapper = shallow(<MultipleCustomHook/>);
        expect(wrapper).toMatchSnapshot();
    });
    
    test('Debe mostrar la informacion que espero', () => {
        // Voy a pasar esto por un mock , solo me improta el resultado
        // useFetch
        useFetch.mockReturnValue({
            data: [{
                author: 'Fernando',
                quote: 'Hola Mundo'
            }],
            loading: false,
            error:null
        });

        const wrapper = shallow(<MultipleCustomHook/>);

        // Como hacer busqueda en el wrapper con el metodo find();
        // quier buscar por clases
        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo');
        expect(wrapper.find('footer').text().trim()).toBe('Fernando');

        // console.log(wrapper.html());



    });
    
})
