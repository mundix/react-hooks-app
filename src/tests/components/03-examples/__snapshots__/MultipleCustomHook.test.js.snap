// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHook/> debe mostrase correctamente 1`] = `
<div>
  <h1>
    Breakikng Bad Quotes!!!
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading ...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente Quote
  </button>
</div>
`;
